docs/notebooks/trust_region.pct.py

```diff
@@ -0,0 +1,286 @@
+
+
+plot_final_result(dataset)
+
+# %% [markdown]
+# We can also visualize the progress of the optimization by plotting the trust regions at each step.
+# The trust regions are shown as translucent boxes, with the current optimum point in each region
+# shown in matching color.
+#
+# Note there is only one trust region in this plot, but the rule in the next section will show multiple trust
+# regions.
+
+# %%
+import base64
+import io
+from typing import List
+
+import imageio
+import IPython
+import matplotlib.pyplot as plt
+
+from trieste.experimental.plotting import plot_trust_region_history_2d
+
+
+def fig_to_frame(fig: plt.Figure) -> np.ndarray:
+    fig.canvas.draw()
+    size_pix = fig.get_size_inches() * fig.dpi
+    image = np.frombuffer(fig.canvas.tostring_rgb(), dtype="uint8")
+    return image.reshape(list(size_pix[::-1].astype(int)) + [3])
+
+
+def frames_to_gif(
+    frames: List[np.ndarray], duration=5000
+) -> IPython.display.HTML:
+    gif_file = io.BytesIO()
+    imageio.mimsave(gif_file, frames, format="gif", loop=0, duration=duration)
+    gif = IPython.display.HTML(
+        '<img src="data:image/gif;base64,{0}"/>'.format(
+            base64.b64encode(gif_file.getvalue()).decode()
+        )
+    )
+    return gif
+
+
+def plot_history(result: trieste.bayesian_optimizer.OptimizationResult) -> None:
+    frames = []
+    for step, hist in enumerate(
+        result.history + [result.final_result.unwrap()]
+    ):
+        fig, _ = plot_trust_region_history_2d(
+            branin,
+            search_space.lower,
+            search_space.upper,
+            hist,
+            num_init=num_initial_data_points,
+        )
+
+        if fig is not None:
+            fig.suptitle(f"step number {step}")
+            frames.append(fig_to_frame(fig))
+            plt.close(fig)
+
+    IPython.display.display(frames_to_gif(frames))
+
+
+plot_history(result)
+
+# %% [markdown]
+# ## Batch trust region rule
+#
+# Next we demonstrate how to run Bayesian optimization with the batch trust region rule.
+#
+# ### Create the batch trust region acquisition rule
+#
+# We achieve Bayesian optimization with trust region by specifying `BatchTrustRegionBox` as the
+# acquisition rule.
+#
+# This rule needs an initial number `num_query_points` of sub-spaces (or trust regions) to be
+# provided and performs optimization in parallel across all these sub-spaces. Each region
+# contributes one query point, resulting in each acquisition step collecting `num_query_points`
+# points overall. As the optimization process continues, the bounds of these sub-spaces are
+# dynamically updated.
+#
+# In addition, this rule requires the specification of a batch aquisition base-rule for performing
+# optimization; for our example we use `EfficientGlobalOptimization` coupled with
+# `ParallelContinuousThompsonSampling`.
+#
+# Note: the number of sub-spaces/regions must match the number of batch query points.
+
+# %%
+num_query_points = 5
+
+init_subspaces = [
```

Comment:
Not related to the notebook in itself, but have you considered doing this by default at initialisation stage within `BatchTrustRegionBox` if `init_subspaces` isn't provided?

Reply:
I think it would be useful to have this initialisation, yes. More in line with the design choices we've made for trieste so far, which is to provide default for many methods so the entry cost for the user is reduced?

Reply:
I have added some explanation for `SingleObjectiveTrustRegionBox` in the notebook.

It will be simpler to add the default initialisation in #778, as that PR already has changes to `init_subspaces` behaviour to allow a single value to be specified (in addition to a list). I'll expand that to include default initialisation.

Reply:
I have added some explanation for `SingleObjectiveTrustRegionBox` in the notebook.

It will be simpler to add the default initialisation in #778, as that PR already has changes to `init_subspaces` behaviour to allow a single value to be specified (in addition to a list). I'll expand that to include default initialisation.